lib/ecto/adapters/sql/sandbox.ex

```diff
@@ -0,0 +1,320 @@
+    {:reply, reply, s}
+  end
+
+  def handle_call({:query, query, log, opts}, from, %{queries: queries} = s) do
+    {:noreply, %{s | queries: :queue.in({query, log, opts, from}, queries)}}
+  end
+
+  ## Mode
+
+  def handle_call(:mode, _, %{mode: mode} = s) do
+    {:reply, mode, s}
+  end
+
+  ## Stop
+
+  def handle_call(:stop, _, s) do
+    {:stop, :normal, :ok, s}
+  end
+
+  ## Cancel
+
+  @doc false
+  def handle_cast({:cancel, ref}, %{ref: ref} = s) do
+    handle_cast({:checkin, ref}, s)
+  end
+  def handle_cast({:cancel, ref}, %{clients: clients} = s) do
+    {:noreply, %{s | clients: :queue.filter(&cancel(&1, ref), clients)}}
+  end
+
+  ## Checkin
+
+  def handle_cast({:checkin, ref}, %{ref: ref} = s) do
+    s = s
+      |> demonitor()
+      |> dequeue()
+    {:noreply, s}
+  end
+
+  ## DOWN
+
+  @doc false
+  def handle_info({:DOWN, mon, _, _, _}, %{monitor: mon, fun: :run} = s) do
+    {:noreply, dequeue(%{s | fun: nil, monitor: nil, ref: nil})}
+  end
+  def handle_info({:DOWN, mon, _, _, _}, %{monitor: mon, mode: :raw} = s) do
+    s = %{s | fun: nil, monitor: nil, ref: nil}
+      |> reset()
+      |> dequeue()
+    {:noreply, s}
+  end
+  def handle_info({:DOWN, mon, _, _, _}, %{monitor: mon, mode: :sandbox} = s) do
+    s = %{s | fun: nil, monitor: nil, ref: nil}
+      |> dequeue()
+    {:noreply, s}
+  end
+  def handle_info({:DOWN, mon, _, _, _}, %{clients: clients} = s) do
+    down = fn({_, _, _, mon2}) -> mon2 !== mon end
+    {:noreply, %{s | clients: :queue.filter(down, clients)}}
+  end
+
+  ## EXIT
+
+  def handle_info({:EXIT, conn, reason}, %{conn: conn} = s) do
+    {:stop, reason, %{s | conn: nil}}
+  end
+
+  ## Info
+
+  def handle_info(_, s) do
+    {:noreply, s}
+  end
+
+  ## Terminate
+
+  @doc false
+  def terminate(_, %{module: module, conn: conn}) do
+    conn && module.disconnect(conn)
+  end
+
+  ## Helpers
+
+  defp checkout(pool, fun, timeout) do
+    ref = make_ref()
+    case :timer.tc(fn() -> do_checkout(pool, ref, fun, timeout) end) do
+      {queue_time, {:ok, mod_conn}} ->
+        {:ok, ref, mod_conn, queue_time}
+      {_, {:error, _} = error} ->
+        error
+    end
+  end
+
+  defp do_checkout(pool, ref, fun, timeout) do
+    try do
+      GenServer.call(pool, {:checkout, ref, fun}, timeout)
+    catch
+      :exit, {:timeout, _} = reason ->
+        GenServer.cast(pool, {:cancel, ref})
+        exit(reason)
+      :exit, {:noproc, _} ->
+        {:error, :noproc}
+    end
+  end
+
+  defp query(pool, query, log, opts, timeout) do
+    GenServer.call(pool, {:query, query, log, opts}, timeout)
+  end
+
+  defp cancel({_, ref, _, mon}, ref) do
+    Process.demonitor(mon, [:flush])
+    false
+  end
+  defp cancel(_, _) do
+    true
+  end
+
+  defp demonitor(%{monitor: mon} = s) do
+    Process.demonitor(mon, [:flush])
+    %{s | fun: nil, ref: nil, monitor: nil}
+  end
+
+  defp dequeue(%{queries: queries} = s) do
+    case :queue.out(queries) do
```

Comment:
Would it make more sense then to still have it behind poolboy? We need to ship with a default queue... so we may as well use it?